client/task_runner.go

```diff
@@ -938,27 +939,40 @@ func (r *TaskRunner) shouldRestart() bool {
 	destroyed := r.destroy
 	r.destroyLock.Unlock()
 	if destroyed {
-		r.logger.Printf("[DEBUG] client: Not restarting task: %v because it has been destroyed due to: %s", r.task.Name, r.destroyEvent.Message)
+		r.logger.Printf("[DEBUG] client: Not restarting task: %v because it has been destroyed", r.task.Name)
 		r.setState(structs.TaskStateDead, r.destroyEvent)
 		return false
 	}
 
 	return true
 }
 
-// killTask kills the running task, storing the reason in the Killing TaskEvent.
-func (r *TaskRunner) killTask(reason string) {
+// killTask kills the running task. A killing event can optionally be passed and
+// this event is used to mark the task as being killed. It provides a means to
+// store extra information.
+func (r *TaskRunner) killTask(killingEvent *structs.TaskEvent) {
 	r.runningLock.Lock()
 	running := r.running
 	r.runningLock.Unlock()
 	if !running {
 		return
 	}
 
-	// Mark that we received the kill event
+	// Get the kill timeout
 	timeout := driver.GetKillTimeout(r.task.KillTimeout, r.config.MaxKillTimeout)
-	r.setState(structs.TaskStateRunning,
-		structs.NewTaskEvent(structs.TaskKilling).SetKillTimeout(timeout).SetKillReason(reason))
+
+	// Build the event
+	var event *structs.TaskEvent
+	if killingEvent != nil {
+		event = killingEvent
+		event.Type = structs.TaskKilling
+	} else {
+		event = structs.NewTaskEvent(structs.TaskKilling)
```

Comment:
[nit] I have a slight preference for forcing the caller to pass this in and assuming the argument is non-nil (as this is slightly safer if people come in and change this code in the future and have to remember that the arg can be nil).